jax_md/elasticity.py

```diff
@@ -0,0 +1,526 @@
+    C = _convert_responses_to_elastic_constants(response_all)
+    
+    #JAX does not allow proper runtime error handling in jitted function. 
+    # Instead, if the user requests a gradient check and the check fails,
+    # we convert C into jnp.nan's. While this doesn't raise an exception,
+    # it at least is very "loud". 
+    if gradient_check is not None:
+      maxgrad = jnp.amax(jnp.abs(grad(energy_fn)(R, **kwargs)))
+      C = lax.cond(maxgrad > gradient_check, lambda _: jnp.nan * C, lambda _: C, None)
+    
+    if check_convergence: 
+      return C, converged
+    else:
+      return C
+
+  return calculate_EMT
+
+
+def _get_mandel_mapping_weight(DIM, dtype):
+  if DIM == 2:
+    m_map  = jnp.array([[0,0],[1,1],[0,1]], dtype=jnp.int8)
+    weight = jnp.array([1,1,jnp.sqrt(2)], dtype=dtype)
+    return m_map, weight
+  elif DIM == 3:
+    m_map  = jnp.array([[0,0],[1,1],[2,2],[1,2],[0,2],[0,1]], dtype=jnp.int8)
+    weight = jnp.array([1,1,1,jnp.sqrt(2),jnp.sqrt(2),jnp.sqrt(2)], dtype=dtype)
+    return m_map, weight
+  else:
+    raise AssertionError('DIM must be 2 or 3')
+
+def tensor_to_mandel(T: Array) -> Array:
+  """ Convert a tensor to Mandel notation
+
+  Mandel notation is a way to represent symmetric second-rank tensors and
+  fourth-rank tensors with minor symmetries in a reduced form. Pairs of indices
+  are combined as follows:
+      
+      For tensors of shape (2,2) or (2,2,2,2):
+      tensor indices        Mandel indices
+      0,0 -------------->   0
+      1,1 -------------->   1
+      0,1 or 1,0 ------->   2
+
+      For tensors of shape (3,3) or (3,3,3,3):
+      tensor indices        Mandel indices
+      0,0 -------------->   0
+      1,1 -------------->   1
+      2,2 -------------->   2
+      1,2 or 2,1 ------->   3
+      0,2 or 2,0 ------->   4
+      0,1 or 1,0 ------->   5
+  
+  If mandel_index(i,j) performs the above index mapping, then the input T and 
+  output M satisfy
+    M[mandel_index(i,j)] = T[i,j] * w
+  or
+    M[mandel_index(i,j), mandel_index(k,l)] = T[i,j,k,l] * w(i,j) * w(k,l)
+  where
+    w(i,j) = 1       if i==j
+           = sqrt(2) if i!=j 
+  is a weight that is used to ensure proper contraction rules. Here (and only 
+  here) we do not assume major symmetries in fourth-rank tensors.
+
+  Args:
+    T: Array with 4 possible shapes:
+      1. T.shape == (2,2)
+         Convert a symmetric array of shape (2,2) to an array of shape (3,)
+      2. T.shape == (3,3)
+         Convert a symmetric array of shape (3,3) to an array of shape (6,)
+      3. T.shape == (2,2,2,2)
+         Convert a tensor of shape (2,2,2,2) with minor symmetries to an array
+         of shape (3,3)
+      4. T.shape == (3,3,3,3)
+         Convert a tensor of shape (3,3,3,3) with minor symmetries to an array
+         of shape (6,6)
+  
+  Output: Array of shape (3,), (6,), (3,3), or (6,6)
+
+  see: https://sbrisard.github.io/janus/mandel.html (accessed 21 April, 2021)
+  """
+  DIM = T.shape[0]
```

Comment:
For consistency with elsewhere can we use `dim` rather than `DIM` (usually all caps is reserved for globals)?

Reply:
Done.